crates/codegen/parser/src/chumsky/terminal_trie.rs

```diff
@@ -30,32 +33,42 @@ impl TerminalTrie {
         }
     }
 
-    fn collect_terminals(&mut self, tree: &CombinatorTree, expression: &ExpressionRef) -> bool {
+    fn collect_terminals(
+        &mut self,
+        tree: &CombinatorTree,
+        expression: &ExpressionRef,
+        inline_references: bool,
+    ) -> bool {
         match &expression.ebnf {
-            EBNF::Choice(exprs) => exprs.iter().all(|e| self.collect_terminals(tree, e)),
-
-            EBNF::DelimitedBy(_)
-            | EBNF::Difference(_)
-            | EBNF::Not(_)
-            | EBNF::OneOrMore(_)
-            | EBNF::Optional(_)
-            | EBNF::Range(_)
-            | EBNF::Repeat(_)
-            | EBNF::SeparatedBy(_)
-            | EBNF::Sequence(_)
-            | EBNF::ZeroOrMore(_) => false,
-
-            EBNF::Reference(name) => {
-                self.collect_terminals(tree, &tree.context.get_tree_by_name(name).expression())
-            }
+            EBNF::Choice(exprs) => exprs.iter().fold(true, |accum, e| {
+                self.collect_terminals(tree, e, inline_references) && accum
+            }),
 
             EBNF::Terminal(string) => {
                 self.0.insert(string.clone(), string.clone());
                 true
             }
+
+            // TODO: only do this if the referenced production is marked as inline
+            EBNF::Reference(name) => {
```

Comment:
I'm adding validation that will disallow using terminals as children of `Rule` productions, and require them to always reference named tokens (that can hold trivia around them).

To make sure I understand this correctly: when that is done, `TerminalTrie` would only be created for `Token` types. Would there be a case where `inline_references` is `false`?

Reply:
Not keen on that idea - I think it complicates the DX and makes the EBNF less clear.

As far as the comment is concerned - that is a remnant of a darker time, when wrong ideas roamed the lands.

Reply:
I'll think about this in the context of the changes I'm currently making to the grammar 👍